### DISCUSSION TITLE
alarm 1 when pressing run or stop on my spindle vfd

### DISCUSSION THREAD

**Discussion by d-granger** [phil-barrett/grblHAL-teensy-4.x#97] [General]
Huanyang 2.2 spindle & vfd.  Spindle and limit cables are shielded & grounded.  Limits are 24v prox sensors which are level shifted via an intermediate optoisolator board mounted next to the 4.1 BOB.  Run, Hold, & E-stop switches are not attached (just have the big red power off button for now).

Alarm 1 triggers as soon as the spindle starts rotating after pressing run.  At first it could not be cleared (retriggered immediately) while the spindle ran, but I added a couple of ferrite beads on each of the spindle phase leads and was able to clear the alarm and command grblHal via ioSender.  The alarm triggers again once the spindle spins down after pressing stop.

My guess is that there is a noise spike when the output stage of the VFD energizes (@ spindle starts rotating) and de-energizes (after ramp down from stop).  Unfortunately, Alarm 1 requires a Reset/Unlock to recover making it pretty difficult to run grbl code that manages the spindle.

So I'm looking for suggestions for hardening my setup.  I have an EMI filter for the input side of the VFD, but some VFD manufacturers caution against them and my last Huanyang blew up after sitting for a couple of hours with the spindle off and the filter in place.  I've tried to isolate the spindle cable and avoid any unnecessary parallel runs, but the cable chains force spindle, stepper, encoder, & limit cables to run along side each other for a couple of meters or so.  At least they are all shielded and the shields are grounded.

So, any ideas or thoughts about places to focus?

**Comment by phil-barrett** [comment]
How close is your VFD to the rest of your electronics?  I have mine mounted outside the electronics box on a piece of 6mm aluminum which is earth grounded and the VFD ground goes to that.   (edit: and no issues)

**Reply by d-granger** [reply]
Good point.  I just got my replacement Huanyang VFD and haven't mounted things permanantly yet.  The controller, drivers, pwr, etc are all mounted to the enclosure back panel but it's laying on the shelf beneath the CNC, and the VFD is laying next to the CNC.  I'll try tidying things up.  I do have another VFD (Durapulse) that supposedly has built in noise supression on both the input and output sides.  I think I'll give it a try first.

**Comment by stevempotter** [comment]
I needed to have my VFD in the same box as all my other electronics because I need everything to be portable (see https://youtu.be/SJ8Mcgfx2aY?si=xaiwjhbplV6Exzpe ). I had the same problems with huge interference from the VFD, when it was on but not running usually. I solved it by using shielded cabling for everything, and careful grounding of the shields.
And also by putting a low-pass filter on the lines that were giving me the most trouble (fan tachometer signal lines). If I recall, I chose a cutoff frequency of 1 kHz for the RC filter. 
I recall asking Phil (somewhere here on Github issues) if he would put RC filters on the next version of the breakout board... Any progress on that?
BTW, I use hall-effect magnetic switches for my endstop sensors, and they are very immune to noise. Never had a false trigger, with a 1k pullup resistor. See https://youtu.be/ztcMPmEYCmw?si=JTf-rY07BjBvFTxi&t=775

**Reply by phil-barrett** [reply]
I don't recall the specific conversation about filters. Went looking for it (here and in the grblHAL discussions and issues sections) but was unsuccessful. Could you remind of which i/o pins?

**Reply by phil-barrett** [reply]
Found it, in email.  Limit inputs.  The optos form an input filter due to their rise and fall times.  Doing rough calculations using 1/2 of typical rise and fall (4 and 7 uS) - we get pulses shorter than 2 uS being rejected which works out to be around 500 KHz for NO switches, 285KHz for NC. A more accurate analysis would figure rise time to min Vih and fall time to max Vil.  

I am unsure if there is a lot of benefit to adding filtering on top of the optos.  It certainly will help in the case of massive EMI though that situation will likely need shielding and other anti-EMI measures regardless of filtering on the limit inputs.

**Reply by d-granger** [reply]
Phil's suggestion was very helpful.  I moved the VFD further from both the controller electronics and the CNC itself and the alarms became far less frequent.  I then added a few more ferrites on the output leads so I had 2 beads on each individual lead and 2 more common to all three.  I also reworked the connection path from the cable shields to ground.  After that, I had no more alarms when starting, running, or stopping the VFD,

To be thorough I swapped the Durapulse VFD for the Huanyang and removed the ferrites.    The Durapulse alarmed whenever I ran the spindle as did the Huanyang.  Adding back the ferrites cured the alarms.  So both VFDs required noise suppression.  I'm now running with the Durapulse primarily because it has a setting to control the fan which defaults to off at temps below 60c.  It's really pleasant to not have a howling fan constantly running.  An added bonus is the Durapulse manual is very detailed and written by someone who apparently speaks English as their first language.

**Reply by phil-barrett** [reply]
The Durapulse VFD sounds like a winner. Based on a quick look, they aren't that much more expensive than the HY inverters.  I find the HY VFD fan to be amazingly loud, close to the spindle - especially when debugging interfaces in the lab where it sits not far from me.
